Convert DG-RePlAce algorithm to Kokkos

kamilrakoczy opened the pull request:
This MR converts [DG-RePlAce](https://github.com/The-OpenROAD-Project/OpenROAD/pull/4959) algorithm that was originally written for CUDA to [Kokkos](https://github.com/kokkos/kokkos).

Kokkos provides abstraction for writing parallel code that can be translated into several backends including CUDA, OpenMP and C++ threads.

Tested on single run with RTX 3090 and i7-8700 CPU @ 3.20GHz using `ariane133` design.

|                             | original placer | CUDA implementation  | Kokkos (CUDA backend) | Kokkos (OpenMP backend) | Kokkos (Threads backend) |
|-----------------------------|------------------------------------|---------------------------------------------|-----------------------|-------------------------|--------------------------|
| ariane133 global place time | 11:27.39                           | 0:57.70                                     | 1:33.49               | 3:24.12                 | 6:08.94                  |

ZhiangWang033 committed:
DGRePlAce

Signed-off-by: ZhiangWang033 <zhw033@ucsd.edu>

kbieganski committed:
Replace CUDA with Kokkos library

Co-authored-by: Kamil Rakoczy <krakoczy@antmicro.com>
Co-authored-by: Jan Bylicki <jbylicki@antmicro.com>

Signed-off-by: Krzysztof Bieganski <kbieganski@antmicro.com>
Signed-off-by: Kamil Rakoczy <krakoczy@antmicro.com>
Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

maliberty commented:
Earlier it was reported the runtime difference to be minimal but 0:57.70 vs 1:33.49 is more substantial.  Is this expected?

kamilrakoczy commented:
> Earlier it was reported the runtime difference to be minimal but 0:57.70 vs 1:33.49 is more substantial. Is this expected?

Earlier measurements were done when some parts was still using native CUDA and using different design (`black-parrot`).
This measurements are single run on local machine while using it for other things too, so they are not very accurate.

I'd expect, it should be possible to achieve similar runtime using Kokkos, This results might suggest, that there are some unnecessary memory copies between host/device, but this needs to be investigated further.

maliberty commented:
What was the thinking behind making kokkos a dependency but kokkos-fft a submodule?  It seems like they could both be build dependencies (and added to the DependencyInstaller with an option).

QuantamHD commented:
> Please try to get a more precise measure of the runtime difference as this is important in deciding whether Kokkos is a good alternative to direct CUDA coding.
> 
> Do all the various versions produce the same result? That is also important.

I think I would say direct CUDA coding isn't really a viable option. I would be personally opposed to its inclusion. I think Kokkos or something like it is the only viable path forward. The runtime differences don't look significant if you compare it to the overall speedup achieved. 

We're going for a pragmatic path forward, and to me this meets my bar for the goals we set out.

maliberty commented:
> I think I would say direct CUDA coding isn't really a viable option. I would be personally opposed to its inclusion. 

You personally pushed for the inclusion of gpuSolver.cu and said its was valuable as a template for future development.  Shall we delete it?  I was never in favor.

A 50% overhead is worth exploring to at least understand if not eliminate.

QuantamHD commented:
> You personally pushed for the inclusion of gpuSolver.cu and said its was valuable as a template for future development. Shall we delete it? I was never in favor.

I think that seems like the right move at this point. With more time and context I don't think it's viable for us to maintain two codebases.

kamilrakoczy commented:
> Do all the various versions produce the same result? That is also important.

No they don't and it was quite surprising, as I expected that original code and Kokkos with CUDA backend will produce the same result.
We investigated this and it turned out that it is because Kokkos passes all files that depends on it through [nvcc_wrapper](https://github.com/kokkos/kokkos/blob/master/bin/nvcc_wrapper). This wrapper converts host compiler options (`g++`) to `nvcc` options and uses `nvcc` to compile all Kokkos-dependent sources. This is done to allow device code in single .cpp file instead of separate .cu file for it.

[NVCC](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#the-cuda-compilation-trajectory) should do pre-processing and compilation for device code and produce CUDA binary and it should leave host code for host compiler.

We checked that when `nvcc` is used to compile InitialPlace, Eigen `solveWithGuess` returns different results with exactly the same inputs comparing to using `g++` directly.

I suspect that this issue isn't only related to Eigen: when I disabled initial placement, runtime of Kokkos and original code were almost the same, but results were still different (I haven't investigated reason for this).

> What was the thinking behind making kokkos a dependency but kokkos-fft a submodule? It seems like they could both be build dependencies (and added to the DependencyInstaller with an option).

kokkos-fft is header only interface library that translates FFT calls into proper backend by detecting enabled backends in Kokkos, but I agree, if preferred, both kokkos and kokkos-fft could be dependencies.

> A 50% overhead is worth exploring to at least understand if not eliminate.

I think this overhead is due to different initial placement, when initial placement is disabled runtime is very similar:

|                                                       | CUDA implementation  | Kokkos (CUDA backend) |
|-------------------------------------------------------|----------------------|-----------------------|
| ariane133 global place time without initial placement | 0:55.52              | 0:58.25               |


I also did precise measurements using RTX 3080, 8 vCPU i9-12900 @ 2.42 GHz and 32GB of RAM with 10 runs using `ariane133` design:

|                          | min time [min] | avg time [min] | med time [min] | max time [min] |
|--------------------------|----------------|----------------|----------------|----------------|
| CUDA implementation      | 0:45           | 0:48           | 0:47           | 0:53           |
| Kokkos (CUDA backend)    | 1:53           | 1:57           | 1:57           | 2:00           |
| Kokkos (OpenMP backend)  | 1:50           | 2:04           | 1:54           | 2:37           |
| Kokkos (threads backend) | 3:42           | 3:43           | 3:43           | 3:45           |

kamilrakoczy commented:
> Is nvcc calling g++ with different flags?

Arguments that are passed to `nvcc` and that `nvcc` should pass to g++ are the same.
I haven't investigated yet how (with what flags) g++ is invoked from `nvcc`.

maliberty commented:
another possibility is that it is invoking a different g++ binary from another path

maliberty commented:
Converted to a draft due to no progress.

jbylicki committed:
Add Kokkos in-tree building and configuration

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

jbylicki commented:
I've rebased this branch onto latest `master` and started resolving the mentioned issues:

- Eigen’s `solveWithGuess()` behaves differently on the Kokkos branch (with a suggestion that this is caused by nvcc_wrapper, a part of Kokkos responsible for redirecting compilations, not pertaining to CUDA, to the host compiler):

I've found that to not be the case. Early, I've recreated the same condition (where Eigen was running slowly) using `clang++` as the Kokkos compiler and I've confirmed that `nvcc_wrapper` was not used then. The problem was Eigen, when detecting `CUDA` availability, was trying to use it. Nevertheless, I saw no peak in GPU usage when `initial_place` was running, so I've disabled it and saw the numbers return to baseline (the same as in the CUDA-native implementation).

- What is the performance difference between Kokkos and CUDA-native implementations?

To prioritize merging of GPU-accelerated placement, the focus was to get the branch issue-free before optimizing. In my testing, Kokkos-based algorithm on `black-parrot` spends about 10 seconds in `libcuda.so`, whereas the CUDA-native implementation spends around 5. All other timings are comparable, making the entire run about 5 seconds longer.

Future / subsequent work:
- Make Kokkos a submodule: Due to varying conditions on host machines, most Kokkos libraries available as a package ship without either CUDA or OMP support. Having a dependency that has to be manually compiled and set correctly to have a functioning and fast implementation might intruduce complexity for the end user. Therefore, I suggest not migrating `kokkos-fft` to be a dependency and using `kokkos`, that is already cloned as a submodule to `kokkos-fft`, as an in-tree library. The issue I'm currently facing is that internal deprecations of CMake symbols are being triggered when Kokkos' compilation is triggered as a child project and not the parent.
- Optimize memory accesses and the Kokkos implementation itself: I've confirmed that memory copying is one of the causes of the algorithm being slower, and fixes are in development, waiting for the more pressing issues to be resolved.

jbylicki commented:
I added a configuration option to `etc/Build.sh`, `-use_gpl2` that will include the `gpl2` subdirectory and launch the compilation of `kokkos` via `kokkos-fft` in CMake. I additionally assigned the `-gpu` flag from the build script to enable the CUDA backend in Kokkos.

maliberty commented:
I would prefer to see kokkos as part of the dependency installer rather than as a submodule.  There should be no need to compile it for each workspace on a machine.

jbylicki commented:
With the current setup, it would be possible to support both compilation schemes, with the priority set towards the `DependencyInstaller` - if a system-wide Kokkos installation would be detected, it will be used during compilation. I would suggest leaving the possibility to use in-tree `Kokkos` and `kokkos-fft` (if `kokkos-fft` was also moved to be downloaded via `DependencyInstaller`), as the script is tailored only towards Ubuntu users. If a system-wide package is not detected, both dependencies can be installed via `FetchContent` and built in-tree.

maliberty commented:
If someone wants to put a local copy in-tree that's fine but I'd like to avoid having a submodule.

jbylicki commented:
I'll add support for `kokkos` and `kokkos-fft` via the `DependencyInstaller` then. The submodule could be deleted while keeping in-tree support - CMake would in case of a system-wide package being absent handle the download by the `FetchContent` directive, and the build would have conditionals in place to link correctly.

jbylicki committed:
Moved Kokkos/Kokkos-fft to be primarly handled via DependencyInstaller

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

sgizler committed:
Force LayoutRight on CUDA

Removes interpretation of LayoutRight data as LayoutLeft

Fixes `input and output extents must be the same except for
the transform axis` gpl2 error caused by incomplete workaround against
differing default 2d data layouts beetween CUDA and CPU.

I considered alternative approches (such as getting rid of LayoutRight
specific code entirely) but they turned out to be unproportionaly complex.

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

sgizler committed:
Replace non-deterministic paralel reduces with serial loops

Replace non-deterministic paralel reduces with serial loops that give
same result regardless of platform

NOTE: This change results in serious performance degradation
Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

sgizler committed:
Always calculate fft on host for consistency

Always calculate fft on host to avoid differing results between impls

NOTE: This may have some performance repercussions.
Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

jbylicki committed:
gpl2: Utilize nested parallelism in computeBCPosNegKernel

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

jbylicki commented:
I've added nested parallelism to the most time consuming kernel - `computeBCPosNegKernel`. After rebasing both branches to the same base commit, the performance results are as follows for the `black-parrot` design with the `CUDA` backend:
- CUDA-native: 24.606 seconds (total time: 114.50 s, skipped intial place: 94.49 s)
- Kokkos:  23.614 seconds (total time: 114.42 s, skipped intial place: 95.07 s)

Additionally, a concern was raised wrt. non-deterministic results that are returned from Kokkos, depending on the compute device used for processing. To validate the flow, each variant was subjected to a run from syntheis to the `final` step. While it's true that those results are varying, they have *minimal* impact on the actual parameters of the finished flow. Additionally, the results are deterministic on a per-device basis, even when the compute device is calculating under heavy external loads (especially applicable for GPUs).

Test subjects were:
- `master` branch commit `7e0fce872123`, as baseline and base for other branches
- `cuda-native`, the original CUDA-native implementation, rebased onto the same base as other branches
- `kokkos-cpu`, the Kokkos-based flow, ran on the `OpenMP` backend
- `kokkos-gpu`, the Kokkos-based flow, ran on the `CUDA` backend

Metrics collected were taken from the final report and log, and were:
- Total Negative Slack (`tns`)
- Worst Negative Slack (`wns`)
- Total power
- Design area and utilization

Results:
| Branch                | TNS   | WNS   | Design area, utilization   | Total Power |
|-----------------------|-------|-------|----------------------------|-------------|
| `master`              | -2.42 | -2.42 | 760397 u^2 45% utilization | 2.57e-01 W  |
| `cuda-native`         | -2.40 | -2.40 | 753511 u^2 44% utilization | 2.49e-01 W  |
| `kokkos-cpu`          | -2.49 | -2.49 | 753608 u^2 44% utilization | 2.50e-01 W  |
| `kokkos-gpu`          | -2.44 | -2.44 | 753674 u^2 44% utilization | 2.50e-01 W  |

sgizler committed:
Calc math funcs in double precision for consistency

Device and host may use different implementations of math functions
giving different results which is not desirable in OpenROAD
The fix relies on (possibly wrong) assumption, that the error of
double precision built-in function is less than precision of float.

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

maliberty commented:
Very nice!  How is the cpu vs gpu runtime with your latest changes?  Is this ready for review?

jbylicki committed:
gpl2: Fix lint violations in implementation

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

jbylicki commented:
Yes, it's ready for review. I've applied the suggested clang-tidy fixes and added the missing RockyLinux9 package.

The performance difference between `CUDA` and `OpenMP` backends on `black_parrot` is:
- `CUDA`: 85.38 s (`dg_global_place` call time: 20.46 s)
- `OpenMP`: 96.58 s (`dg_global_place` call time: 29.83 s)

The test setup is an Intel i7-8700 and a NVIDIA GTX 1080Ti

maliberty commented:
What is the current gpl time on that design & system?

maliberty commented on src/gpl2/include/gpl2/DgReplace.h:
```diff
@@ -0,0 +1,212 @@
+            utl::Logger* logger);
+  void reset();
+
+  // The three main functions
+  void doInitialPlace();
+  int doNesterovPlace(int start_iter = 0);
```
I only see two

jbylicki replied:
Fixed the comment, it was here since the CUDA-native implementation, there's no clear indication what would've been the third function.

maliberty commented on src/gpl2/include/gpl2/DgReplace.h:
```diff
@@ -0,0 +1,212 @@
+  grt::GlobalRouter* fr_;
+  utl::Logger* log_;
+
+  // We should only have one placerBaseCommon, timingBase and routeBase
+  // But we need multiple placerBases to handle fences and multiple domains
```
This comment seems to have become separate from its context (move down 4 lines)

jbylicki replied:
Done

maliberty commented on src/gpl2/src/dct.cpp:
Did you author this code?

kbieganski replied:
It was originally written by @ZhiangWang033 (as can be seen in the commit history), then ported to Kokkos by the other committers (including me).

maliberty commented on src/gpl2/src/densityOp.h:
```diff
@@ -0,0 +1,128 @@
+  int coreUx_;
+  int coreUy_;
+
+  // We need to store all the statictis information for each bin
```
typo: statictis

jbylicki replied:
Corrected

maliberty reviewed (changes requested):
Partial review

jbylicki commented on etc/Build.sh:
```diff
@@ -121,6 +122,9 @@ while [ "$#" -gt 0 ]; do
             echo "${1} requires an argument" >&2
             _help
             ;;
+        -use_gpl2)
```
Done

jbylicki commented on etc/DependencyInstaller.sh:
```diff
@@ -76,6 +76,7 @@ _installCommonDev() {
     gtestChecksum="a1279c6fb5bf7d4a5e0d0b2a4adb39ac"
     bisonVersion=3.8.2
     bisonChecksum="1e541a097cda9eca675d29dd2832921f"
+    kokkosfftVersion="2c616d29a7ad0c390259efeb9224115bfa6910fd"
```
It accomplishes the same goal in a single defintion and eliminates the need to hash the directory. If it's preferred to keep convention, I can change it.

maliberty replied:
It does not as there is no easy way to know what release (if any) we are using from a commit id.

jbylicki replied:
Changed to a version tag. Other dependencies managed by git were not getting hashed either, as it would require to `tar` the directory.

jbylicki commented on etc/DependencyInstaller.sh:
```diff
@@ -173,6 +174,25 @@ _installCommonDev() {
+        git submodule update --recursive
+        ARGS_KOKKOSFFT=" -DCMAKE_BUILD_TYPE=Release -DKokkos_ENABLE_OPENMP=ON -DKokkosFFT_ENABLE_INTERNAL_KOKKOS=ON"
+        if [[ -n $(which nvcc) ]]; then
+            # Older version of g++ is needed for compatibility with NVCC
+            ARGS_KOKKOSFFT+=" -DKokkos_ENABLE_CUDA=ON -DCMAKE_CXX_COMPILER=g++-10"
```
https://github.com/NVlabs/instant-ngp/issues/119 presents the same issue as I have encountered. Ubuntu is using the supposedly supported gcc 11.4, but it does not compile Kokkos properly.

jbylicki commented on etc/DependencyInstaller.sh:
```diff
@@ -469,36 +497,42 @@ _installOpenSusePackages() {
-        tcllib \
-        wget \
-        zlib-devel
+    zypper -n --gpg-auto-import-keys \
+        --plus-repo https://download.opensuse.org/repositories/games/$(grep ^VERSION= /etc/os-release | sed  's/VERSION="\(.*\)"/\1/g') \
+        --plus-repo https://download.opensuse.org/repositories/Education/$(grep ^VERSION= /etc/os-release | sed  's/VERSION="\(.*\)"/\1/g')\
```
`libfmt11` is only in `games` and `fftw3-devel` in `Education`.

jbylicki commented on src/CMakeLists.txt:
```diff
@@ -270,6 +269,38 @@ endif()
+      set(Kokkos_ENABLE_CUDA ON CACHE BOOL "Set Kokkos CUDA backend" FORCE)
+      set(Kokkos_ENABLE_CUDA_LAMBDA ON CACHE BOOL "Enable Kokkos CUDA Lambda support" FORCE)
+      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DEIGEN_NO_CUDA")
+      if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND NOT CMAKE_CXX_COMPILER_VERSION VERSION_LESS "14")
+        message(FATAL_ERROR "Detected version of GCC is not supported by NVCC. Consider using Clang or disabling GPU support.")
```
It's actually an error on >= 14, but I haven't found a `VERSION_GREATER` or any similar directive. The others are at least theoretically supported according to NVIDIA docs (but as mentioned, there's a g++-11 problem), 14 outright isn't for now.

jbylicki commented on src/Main.cc:
```diff
@@ -453,6 +460,9 @@ static int tclAppInit(int& argc,
 #endif
 
+#ifdef ENABLE_KOKKOS
+    Kokkos::initialize();
```
Done

sgizler committed:
Remove sneaky fp sum in hpwl & parallelize it again

computeWeightedHPWL() suffered from implicit lossy (in case of big
numbers) conversion of int64_t to float.

After fixing it, the summation can be made parallel without introducing
inconsistencies between kokkos configurations.

NOTE: the computeHPWL() never suffered from this issue, but I had
excesivly deparalelized it before.

In this fix, I added safe-guards to both computeWeightedHPWL() and
computeHPWL() for consistency.

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

sgizler committed:
Execute serial reduces on CPU rather than GPU

Serial code is order of magnitude slower to execute on GPU than on CPU

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

sgizler committed:
Replace two reduntant views with single one

Replace dNetWith and dNetHeight with single dNetWidthPlusHeight
The time improvement is small (it could be measurement error as well)

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

sgizler committed:
Use fabs() rather than handmade tenary

I'm a bit suprised, but this simple change reduced time from 2m38s to 2m30s

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

sgizler committed:
Split getDistance to parallel and serial part

Calculate individual distances in parallel, then sum them serially

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

sgizler committed:
Make summations autovectorizable

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

sgizler committed:
Sum X and Y gradients in one go

Before, we executed serial reduction for X and Y separatly.
Now we have parallel calculation of view with abs(X)+abs(Y), and one
serial reduction of it.

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

jbylicki committed:
gpl2: Remove redundant LICENSE file

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

jbylicki commented on src/gpl2/LICENSE:
Removed

jbylicki commented on src/CMakeLists.txt:
```diff
@@ -196,7 +196,6 @@ add_compile_options(
   -Wredundant-decls
   -Wformat-security
   -Wno-sign-compare
-  -Wp,-D_GLIBCXX_ASSERTIONS
```
I'm not certain why it was moved, I've tested with the option defined and it works well, I've restored the change

jbylicki commented:
Currently, the mainline `gpl` runs the `global_place` call for 653.90s with the total run time being 729.30s

jbylicki committed:
gpl2: Make computeBCPosNegKernel deterministic on different compute backends

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

sgizler committed:
Reuse fft plan

NOTE: I used static vars for storing plans. While simple and convienient,
it assumes that N and M won't change between calls (changing them would
result in runtime error)

Signed-off-by: Szymon Gizler <sgizler@antmicro.com>

jbylicki committed:
gpl2: Move unnecessary operations out of loop bodies

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

mikesinouye commented:
Hi @sgizler, is this PR ready for review? It is my understanding that the latest status of this PR is that the performance is consistent with the native CUDA implementation but there is a small amount of variability run-to-run. I discussed with @maliberty that if the amount of PPA variability is low, that is acceptable and expected. Please let me know if that status is correct.

maliberty commented:
Small differences due to numerics are expected

sgizler commented:
There is no variability run-to-run if you execute it on same machine.

Before, there was some variability when you changed the number of threads, or enabled/disabled CUDA. In the worst case I checked (nangate45/ariane133), variability fix incurred ~18% time penalty on CUDA-enabled run, while speeding up non-CUDA run about 14%. On other designs like black parrot the difference was about 3%.

There might still be small variations when moving to a entirely different CPU (i.e not supporting some newer SIMD instructions, or entirely different arch) due to underlying FFT library choosing implementation fine-tuned for certain CPUs.

I think the PR is ready for review.

maliberty commented:
I'll take a look.  Please address formatting & tcl-lint.

maliberty commented:
@vvbandeira please review the dependency install & cmake related changes

vvbandeira commented:
If the `gpl2` compilation is optional, so should the installation of its dependencies. Especially since the script will install NVIDIA packages and some users might have AMD graphics, GPU support on Linux does not have the reputation of being robust, to my knowledge. Some checks in this regard would be nice.

jbylicki committed:
Main.cc: Handle -thread argument correctly for Kokkos

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

jbylicki committed:
Main.cc: gpl2: Re-formatted the changes in accordance with new style guides

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

maliberty commented:
@kamilrakoczy any update?  The last comments are unanswered

jbylicki committed:
etc: DependencyInstaller: Added verbose gpu support and gpl2 compilation

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

jbylicki committed:
gpl2: MakeDgReplace: Switch GPL2 away from sta's TclEncode

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

jbylicki commented:
The branch was rebased onto the latest changes and initialization was fixed to be in compliance with `TclEncode` changes. Both formattting problems should be addressed, but CI is waiting workflow approval to verify it. `DependencyInstaller` was changed to have more verbose GPU and GPL2-related flags.

jbylicki committed:
Main: Added Kokkos finalize hooks cleaning the OR context

Signed-off-by: Jan Bylicki <jbylicki@antmicro.com>

hzeller commented:
Looks like we soon also need bring kokkos and fftw into the [BCR](https://registry.bazel.build/) to get it nicely compiled in the new [bazel build](https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/BUILD.bazel).

jbylicki commented:
@mikesinouye Yes, it is

maliberty commented:
I tried to build this but I get:
```
/home/matt/OpenROAD/src/gpl2/src/dct.cpp: In function ‘void dct_2d_fft(int, int, const Kokkos::View<const Kokkos::complex<float>*>&, const Kokkos::View<const Kokkos::complex<float>*>&, const Kokkos::View<const float*>&, const Kokkos::View<float*>&, const Kokkos::View<Kokkos::complex<float>*>&, const Kokkos::View<float*>&)’:
/home/matt/OpenROAD/src/gpl2/src/dct.cpp:112:21: error: ‘Plan’ in namespace ‘KokkosFFT’ does not name a type
  112 |   static KokkosFFT::Plan fftplan(hostSpace,
```
Have you run into this?

sombraSoft commented on etc/DependencyInstaller.sh:
```diff
@@ -613,6 +658,20 @@ _installDebianPackages() {
         wget \
         zlib1g-dev
 
+    if [[ ${gpuDeps} == "nvidia" ]]; then
+        RELEASE_CODENAME=$(lsb_release -c | awk '{print $2}')
+
+        NEW_LINES="deb http://deb.debian.org/debian/ $RELEASE_CODENAME main contrib non-free
+        deb-src http://deb.debian.org/debian/ $RELEASE_CODENAME main contrib non-free"
+
+        if ! grep -q "$NEW_LINES" /etc/apt/sources.list; then
+            echo "$NEW_LINES" | tee -a /etc/apt/sources.list > /dev/null
+        fi
+        apt-get update
+        apt-get -y install --no-install-recommends libcu++-dev nvidia-cuda-toolkit
+    fi
+
+
```
Ok, there are some issues with this setup for installing `nvidia-cuda-toolkit`.

1. The command `lsb_release -c` on Ubuntu returns an Ubuntu codename like `jammy`, `focal`. The script then tries this codename to access a URL on the debian server `http://deb.debian.org/debian/jammy` which does not exist.
2. Even if the url existed, installing packages from a Debian repository on an Ubuntu system is highly discouraged. Ubuntu is based on Debian but has its own version of libraries and packages. Mixing them can cause critical dependency conflicts that can break the package manager and other essential system services.

I'm currently running Debian 12 bookworm and indeed it needs to have `non-free` in order to find `nvidia-cuda-toolkit`. For ubuntu though, they call it `multiverse` instead of `non-free`.
Also, `libcu++-dev` should already be included with `nvidia-cuda-toolkit` package.

sgizler committed:
Pin Kokkos version to tag rather than origin/main

sgizler committed:
Remove useless dependency (fmtlib)

Neither kokkos, nor gpl2 itself requires it. I guess that it was used
for some debugging in past, and can be safely dropped.

Besides, in other OpenROAD modules we already use spdlog's format
(which is proxy to either std::format or bundled copy of fmtlib)

sgizler committed:
Define missing kokkos flags in DependencyInstaller

Build it with same flags as in cmake

sgizler committed:
Replace deprecated KokkosFFT_ENABLE_HOST_AND_DEVICE

sgizler commented:
> I tried to build this but I get:
> 
> ```
> /home/matt/OpenROAD/src/gpl2/src/dct.cpp: In function ‘void dct_2d_fft(int, int, const Kokkos::View<const Kokkos::complex<float>*>&, const Kokkos::View<const Kokkos::complex<float>*>&, const Kokkos::View<const float*>&, const Kokkos::View<float*>&, const Kokkos::View<Kokkos::complex<float>*>&, const Kokkos::View<float*>&)’:
> /home/matt/OpenROAD/src/gpl2/src/dct.cpp:112:21: error: ‘Plan’ in namespace ‘KokkosFFT’ does not name a type
>   112 |   static KokkosFFT::Plan fftplan(hostSpace,
> ```
> 
> Have you run into this?

I suspect that you have an older version of KokkosFFT (before introduction of `KokkosFFT::Plan`) installed in your system, and that it is being picked up. The quick workaround would probably be to remove `/usr/local/include/kokkos` and rerun `DependencyInstaller.sh`.  We could add some kind of version check to the script, but maybe dropping Kokkos from DependencyInstaller and building it solely in-tree would be better solution? 

The current setup is very complex:
- [DependencyInstaller.sh](https://github.com/The-OpenROAD-Project/OpenROAD/blob/be2d8bddc8a3eb9b6dca602c21c72319e48c8be7/etc/DependencyInstaller.sh#L177) either:
   - detects that KokkosFFT is present on host in which case it does nothing (without any version check)
   - if not present, builds and installs KokkosFFT system-wide
- [CMakeLists.txt](https://github.com/The-OpenROAD-Project/OpenROAD/blob/be2d8bddc8a3eb9b6dca602c21c72319e48c8be7/src/CMakeLists.txt) either:
   - detects that KokkosFFT is present on host in which case it picks it up (not necessarily the one from `DependencyInstaller.sh`)
   - if not present, builds and installs KokkosFFT in-tree

The version mismatch is not the only issue with using `DependencyInstaller.sh` for KokkosFFT installation:
- If you want to switch between using CPU and GPU you have to remove the old installation of Kokkos, rerun DependencyInstaller with proper args and rebuild OpenROAD.
- Kokkos seems to make an assumption that you use the same compiler suite for building the library and for the code using it. This makes it complicated to build OpenROAD with clang.
- Kokkos has to be built with non-default compile-time options in order to give deterministic results. We would have to add some extra (potentially complicated) checks for that too.
- Even if we add version and flags checks in `DependencyInstaller.sh`, there is no guarantee that CMake will pick the correct instance of Kokkos if user already had it installed globally

Building Kokkos solely in-tree would solve these problems and decrease maintenance burden related to having two independent methods of installing Kokkos.

maliberty commented:
I think we should finish this off

hzeller commented:
If this continues, I can work on making kokkos build with bazel (and once it works also put on BCR)

maliberty commented:
That would be great.  I've been holding off until we can retire cmake to avoid having to deal with it.

dev-minjae commented:
Late-comer thought — I've spent time reviewing gpl and have an RTX 5090 on hand. After reading both gpl and gpl2, much of gpl2 (Nesterov body, WA wirelength, BiCGSTAB) looks like the same algorithms as gpl. So instead of a separate gpl2 library, you could keep gpl and add Kokkos kernels behind an optional build flag, one function per PR. The CMake side can reuse -DGPU=ON in etc/Build.sh now; the bazel side will work once @hzeller finishes putting Kokkos on BCR. First PR could be getHpwl to set up the option and Kokkos dependency. WA gradient and a Poisson solver using Kokkos-FFT would follow. @sgizler's determinism fixes in this branch apply to either approach.

I'm raising this because the dual-codebase concern keeps coming up, and this approach might address it. Happy to try the first PR if it's useful, or to drop it and let #5352 continue as planned.

maliberty commented:
The goal would be to have a single gpl.  This is a separate code base as that is how it was developed academically but not the preferred end state.  I am glad to take PRs for incrementally adding these ideas directly to gpl/.  Dealing with Kokkos itself will be a significant first step.

dev-minjae commented:
Thanks @maliberty — that's clarifying. I'll put together a first PR with the CMake option and a Kokkos-backed getHpwl, leaving the CPU path as default. If the build-system side needs to align with @hzeller's BCR work, happy to coordinate before posting.

maliberty commented:
Replaced by #10511
